app/layersmodel.cpp

```diff
@@ -13,126 +7,35 @@
  *                                                                         *
  ***************************************************************************/
 
-#include "layersmodel.h"
-#include "qgsquickmapsettings.h"
-
-#include <qgslayertreemodel.h>
-#include <qgslayertreenode.h>
-#include <qgslayertree.h>
-#include <qgsvectorlayer.h>
-#include <qgslayertreemodellegendnode.h>
-#include <qgsproject.h>
-#include "qgsvectorlayer.h"
-#include <qgsmaplayer.h>
-#include <qgswkbtypes.h>
-
-#include <QString>
-
-LayersModel::LayersModel( QObject *parent )
-  : QAbstractListModel( parent )
-{
-}
-
-LayersModel::~LayersModel()
-{
-}
-
-
-void LayersModel::reloadLayers( QgsProject *project )
-{
-
-  if ( !project ) return;
-
-  QgsLayerTreeGroup *root = project->layerTreeRoot();
-
-  // Get list of all visible and valid layers in the project
-  QList< QgsMapLayer * > allLayers;
-  foreach ( QgsLayerTreeLayer *nodeLayer, root->findLayers() )
-  {
-    if ( nodeLayer->isVisible() )
-    {
-      QgsMapLayer *layer = nodeLayer->layer();
-      if ( layer->isValid() )
-      {
-        allLayers << layer;
-        qDebug() << "Found layer: " << layer->name();
-      }
-    }
-  }
-  if ( mLayers != allLayers )
-  {
-    beginResetModel();
-    setLayers( allLayers );
-    endResetModel();
-  }
-}
 
-QgsQuickMapSettings *LayersModel::mapSettings() const
-{
-  return mMapSettings;
-}
+#include "layersmodel.h"
 
-void LayersModel::setMapSettings( QgsQuickMapSettings *mapSettings )
+LayersModel::LayersModel()
 {
-  if ( mapSettings != mMapSettings )
-  {
-    mMapSettings = mapSettings;
-    setLayers( mLayers );
-    emit mapSettingsChanged();
-  }
 }
 
-void LayersModel::updateActiveLayer( const QString &name )
+QVariant LayersModel::data( const QModelIndex &index, int role ) const
 {
-  int row = rowAccordingName( name, firstWritableLayerIndex() );
-  setActiveIndex( row );
-}
+  if ( !index.isValid() )
+    return QVariant();
 
-int LayersModel::activeIndex() const
-{
-  return mActiveIndex;
-}
+  if ( role < LayerNameRole ) // if requested role from parent
+    return QgsMapLayerModel::data( index, role );
 
-QString LayersModel::setActiveIndex( int activeIndex )
-{
-  if ( mActiveIndex != activeIndex )
-  {
-    mActiveIndex = activeIndex;
-  }
-  emit activeIndexChanged();
-
-  QgsMapLayer *layer = activeLayer();
-  if ( layer )
-    return layer->name();
-  else
-    return QString();
-}
+  QgsVectorLayer *vectorLayer;
+  QgsMapLayer *layer;
 
-QVariant LayersModel::data( const QModelIndex &index, int role ) const
-{
-  int row = index.row();
-  if ( row < 0 || row >= mLayers.count() )
-    return QVariant( "" );
+  layer = layerFromIndex( index );
+  if ( !layer || !layer->isValid() ) return QVariant();
 
-  QgsMapLayer *layer = mLayers.at( row );
+  vectorLayer = qobject_cast<QgsVectorLayer *>( layer );
```

Comment:
shall we check for null pointer in case `layer` is not QgsVectorLayer ?

Reply:
Well, yes and no. At first I had a check there, but it is not always true that when `vectorLayer` is null, we want to return nothing. 
There is a case when Raster layer looks for icon source and it would then return empty qvariant. That is why I moved the check to the cases in switch that use vectorLayer.

Reply:
Ah ok I see - I thought there was a chance of null pointer dereference... but looking again, there isn't :-)